File: Python/testing_windows/accumulating_fired_panes.py
#  Copyright 2022 Google LLC
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

import logging
from apache_beam import DoFn
from apache_beam import GroupByKey
from apache_beam import Map
from apache_beam import WindowInto
from apache_beam.options.pipeline_options import PipelineOptions
from apache_beam.testing.test_pipeline import TestPipeline
from apache_beam.testing.test_stream import TestStream
from apache_beam.transforms.trigger import AccumulationMode
from apache_beam.transforms.window import FixedWindows
from apache_beam.transforms.window import TimestampedValue
from apache_beam.utils.windowed_value import PaneInfoTiming


def get_input_stream():
  stream = (
      # timestamp = seconds since epoch
      TestStream().add_elements([
          TimestampedValue("On-Time-1", timestamp=0),
          TimestampedValue("On-Time-2", timestamp=4),
          TimestampedValue("On-Time-3", timestamp=6)
      ])
      .advance_watermark_to(6)  # move watermark to epoch sec 6
      .advance_processing_time(6)  # move processed time to epoch sec 6
      .add_elements([
          # Elements arrive out of order but still within window
          TimestampedValue("On-Time-4", timestamp=4),
          TimestampedValue("On-Time-5", timestamp=7)
      ])
      .advance_watermark_to(11)
      .advance_processing_time(5)
      # processing time crosses window, pane triggered containing 5 elements
      .add_elements([
          TimestampedValue("Late-1", timestamp=0),
          TimestampedValue("Late-2", timestamp=5),
          TimestampedValue("Late-3", timestamp=6),
          TimestampedValue("Late-4", timestamp=7)
      ])
      .advance_watermark_to(15)
      .advance_processing_time(4)
      # processing time within allowed lateness window, pane triggered again
      .add_elements([
          TimestampedValue("Late-5", timestamp=9),
          TimestampedValue("Late-6", timestamp=9),
      ])
      .advance_watermark_to(16)
      .advance_processing_time(1)
      # processing time moves from 14 -> 16, 14 still within lateness window
      .add_elements([
          TimestampedValue("Outside-allowed-lateness", timestamp=9),
          TimestampedValue("On-Time-New-Window-1", timestamp=17)
      ])
      # processing time crossed allowed lateness, first element not triggered
      # On-Time-New-Window-1 shows this process will continue until stopped
      .advance_watermark_to_infinity()
  )

  return stream

def print_with_info(element, window=DoFn.WindowParam, pane=DoFn.PaneInfoParam):
  """Helper function to print window and pane info"""
  print(f"Element: {element[1]} \n\t Window: {window} \n\t Pane Timing: {PaneInfoTiming.to_string(pane.timing)}")


def run():
  options = PipelineOptions(streaming=True)

  p = TestPipeline(options=options)

  window_size_seconds = 10
  window_allowed_lateness_seconds = 5

  stream = get_input_stream()

  (p | stream
     | "Assign elements within pane same key" >> Map(lambda e: ("key", e))
     | "Window into FixedWindows" >> WindowInto(
          # We are using the default triggering which triggers when watermark
          # crosses the window end and whenever valid late element(s) are added
          FixedWindows(size=window_size_seconds),
          allowed_lateness=window_allowed_lateness_seconds,
          # Late elements are combined with previous elements within the window
          accumulation_mode=AccumulationMode.ACCUMULATING)
     | "Combine elements within pane by key" >> GroupByKey()
     | Map(print_with_info)
   )

  p.run()


if __name__ == "__main__":
  logging.getLogger().setLevel(logging.INFO)
  run()

"""
EXPLANATION
1 - Elements "On-Time-X" arrive to the pipeline before window closes
2 - Elements "Late-(1 to 4)" arrive after window closing, but within the allowed late time
3 - Elements "Late-(5, 6)" also arrive after window closing, but within the allowed late time
4 - Element "Outside-allowed-lateness" is discarded as it's outside the allowed lateness


The first trigger contains elements (1), since they arrive before window closing
The second trigger contains elements (2) and elements (1), since we are accumulating
The third trigger contains only elements (3) but also (2) and (1) because of the accumulation
"""

File: Java/src/main/java/testingwindows/AccumulatingFiredPanes.java
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package testingwindows;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.coders.StringUtf8Coder;
import org.apache.beam.sdk.options.PipelineOptions;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.testing.TestStream;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.GroupByKey;
import org.apache.beam.sdk.transforms.ParDo;
import org.apache.beam.sdk.transforms.windowing.FixedWindows;
import org.apache.beam.sdk.transforms.windowing.Window;
import org.apache.beam.sdk.values.KV;
import org.apache.beam.sdk.values.TimestampedValue;
import org.joda.time.Duration;
import org.joda.time.Instant;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class AccumulatingFiredPanes {
    private static final Logger LOG = LoggerFactory.getLogger(AccumulatingFiredPanes.class);

    public static void main(String[] args) {

        PipelineOptions options = PipelineOptionsFactory.fromArgs(args).withValidation().as(PipelineOptions.class);

        Pipeline p = Pipeline.create(options);

        Integer windowLength = 10;
        Integer allowLateSize = 5;

        // Behaviour is explained in detail at the end of the file
        // Note: all elements would fall into the same window
        TestStream<String> streamEvents = TestStream.create(StringUtf8Coder.of())
                .addElements(
                        TimestampedValue.of("On-time-1", new Instant(0)),
                        TimestampedValue.of("On-time-2", new Instant(4000)), // Value of Instant is in millis
                        TimestampedValue.of("On-time-3", new Instant(6000))

                )
                .advanceWatermarkTo(new Instant(6000))// move Watermark time to sec 6
                .advanceProcessingTime(Duration.standardSeconds(6)) // move Processing time to sec 6
                .addElements(
                        TimestampedValue.of("On-time-4", new Instant(4000)),
                        TimestampedValue.of("On-time-5", new Instant(7000))
                )
                .advanceWatermarkTo(new Instant(11000))// move Watermark time to sec 11
                .advanceProcessingTime(Duration.standardSeconds(5)) // move Processing time to sec 11
                .addElements(
                        // element arrives at sec 11 but fall in the window
                        TimestampedValue.of("Late-1", new Instant(0)),
                        TimestampedValue.of("Late-2", new Instant(5000)),
                        TimestampedValue.of("Late-3", new Instant(6000)),
                        TimestampedValue.of("Late-4", new Instant(7000))
                )
                .advanceWatermarkTo(new Instant(14000))
                .advanceProcessingTime(Duration.standardSeconds(3)) // move Processing time to sec 14
                .addElements(
                        TimestampedValue.of("Late-5", new Instant(9000)),
                        TimestampedValue.of("Late-6", new Instant(9000))

                )
                .advanceWatermarkTo(new Instant(16000))
                .advanceProcessingTime(Duration.standardSeconds(2)) // move Processing time to sec 14
                .addElements(
                        TimestampedValue.of("Outside-allowed-lateness", new Instant(9000))

                )
                .advanceWatermarkToInfinity();

        p.apply(streamEvents)
                .apply("KVs", ParDo.of(new DoFn<String, KV<String, String>>() {
                    @ProcessElement
                    public void processElement(ProcessContext c) {
                        c.output(KV.of("key", c.element()));
                    }
                }))
                .apply(Window.<KV<String, String>>into(
                        FixedWindows.of(Duration.standardSeconds(windowLength)))
                        .withAllowedLateness(Duration.standardSeconds(allowLateSize))
                        .accumulatingFiredPanes() // New fired elements would be merged with previous fired elements

                )
                .apply(GroupByKey.create())
                .apply("Log", ParDo.of(new DoFn<KV<String, Iterable<String>>, String>() {
                    @ProcessElement
                    public void processElement(ProcessContext c) {
                        LOG.info("\n TRIGGER " + c.element().getValue().toString());
                        c.output(c.pane().toString());
                    }
                }));


        p.run();
    }
}
        /*
        EXPLANATION
        1 - Elements On-Time-X arrive to the pipeline before window closes
        2 - Elements "Late-(1 to 4)" arrive after window closing, but within the allowed late time
        3 - Elements "Late-(5, 6)" also arrive after window closing, but within the allowed late time
        4 - Element "Outside-allowed-lateness" is discarded as it's outside the allowed lateness


        The first trigger contains elements (1), since they arrive before window closing
        The second trigger contains elements (2) and elements (1), since we are accumulating
        The third trigger contains only elements (3) but also (2) and (1) because of the accumulation
         */
